kernel/vmbytegen.ml

```diff
@@ -287,24 +291,33 @@ let pos_rel i r sz =
         r.in_env := push_fv db env;
         Kenvacc(r.offset + pos)
 
-let pos_universe_var i r sz =
+let pos_instance r sz =
```

Comment:
There is a missing `set_max_stack_size cenv sz` here.

It was already missing in the old code, but I think it was innocuous because we were off by at most one in practice. Now that we do computations to produce an instance block (which in turn might call `pos_instance` repeatedly), I think the maximum stack size might end up being off by more than the safety margin without the call.

Reply:
In fact, `pos_instance` itself does not put anything on the stack, so the call to `set_max_stack_size` does not need to be there. It can and should instead be in `compile_instance::comp_univ`.

(Actually, it should really be in `comp_args`, since this is the function that does an explicit `Kpush`.)

Reply:
I *think* we should have a automatic way to insert this based on some annotations on the stack size needed by each bytecode instruction. But well, in the meantime, this should be fixed indeed.

Reply:
No need, this would be overly complicated and possibly much more error-prone that the current approach. The bytecode compiler is already tracking very accurately the stack size, since this size is critically used to access the arguments of a function (and thus we have a very high confidence in its correctness). We just need to be more systematic in deciding who is responsible for calling `set_max_stack_size`. If we go for stack producers, then `comp_args` should do it. If we go for stack consumers, then `compile_instance` should do it.

Reply:
OCaml's own bytecode compiler is much simpler in this regard. If we ignore the changes due to OCaml 5's effects, there is only one single place where it calls `set_max_stack_size`, which is the start of `compile_lam` (actually, there is a second place, but it is hardly relevant for Coq). Indeed, every place of OCaml's compiler necessarily ends up calling `compile_lam`, including `comp_args`. But Coq's bytecode compiler is trying to be a bit fancy there, and instead of having `comp_args` always call `compile_lam`, it instead calls a closure passed as an argument. Most of the time, this closure is actually `compile_lam`, so we are back to the original design. But in the case of your pull request, you are passing to `comp_args` a brand new function `comp_univ`, which does not perform any tracking, hence the fresh bug.

Reply:
I've added the stack check in comp_instance, in any case.